File: ticketProject10-WebClient/TS-REST - CLIENT/TicketNetworking/src/main/java/ticket/network/protobuffprotocol/ProtoTSWorker.java
package ticket.network.protobuffprotocol;

import services.ServicesException;
import services.TicketSystemObserver;
import services.TicketSystemServices;
import ticket.model.ArtisticRepresentation;
import ticket.model.Buyer;
import ticket.model.Seller;
import ticket.network.dto.ArtisticRepresentationDTO;
import ticket.network.dto.DTOUtils;
import ticket.network.rpcprotocol.Response;
import ticket.network.rpcprotocol.ResponseType;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.rmi.RemoteException;
import java.sql.SQLException;


public class ProtoTSWorker implements Runnable, TicketSystemObserver {
    private TicketSystemServices server;
     private Socket connection;

     private InputStream input;
     private OutputStream output;
     private volatile boolean connected;
     public ProtoTSWorker(TicketSystemServices server, Socket connection) {
         this.server = server;
         this.connection = connection;
         try{
             output = connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
             input = connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
             connected = true;
         } catch (IOException e) {
             e.printStackTrace();
         }
     }

     public void run() {
         while(connected){
             try {
                // Object request=input.readObject();
                 System.out.println("Waiting requests ...");
                 TSProtobufs.TSRequest request = TSProtobufs.TSRequest.parseDelimitedFrom(input);
                 System.out.println("Request received: "+request);
                 TSProtobufs.TSResponse response = handleRequest(request);
                 if (response!=null){
                    sendResponse(response);
                 }
             } catch (IOException e) {
                 e.printStackTrace();
             }
             try {
                 Thread.sleep(1000);
             } catch (InterruptedException e) {
                 e.printStackTrace();
             }
         }
         try {
             input.close();
             output.close();
             connection.close();
         } catch (IOException e) {
             System.out.println("Error "+e);
         }
     }

    public void boughtTicket(ArtisticRepresentation artisticRepresentation) throws ServicesException {
        System.out.println("Someone bought tickets " + artisticRepresentation);
        try {
            sendResponse(ProtoUtils.createBoughtTicketResponse(artisticRepresentation));
        } catch (IOException e) {
            throw new ServicesException("Sending error: " + e);
        }
    }

     private TSProtobufs.TSResponse handleRequest(TSProtobufs.TSRequest request){
         TSProtobufs.TSResponse response = null;
         switch (request.getType()){
             case Login:{
                 System.out.println("Login request ...");
                 Seller seller = ProtoUtils.getSeller(request);
                 try {
                     Seller seller1 = server.login(seller, this);
                     return ProtoUtils.createLoginResponse(seller1);
                 } catch (ServicesException | RemoteException e) {
                     connected=false;
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case Logout:{
                 System.out.println("Logout request");
                 Seller seller = ProtoUtils.getSeller(request);
                 try {
                     server.logout(seller, this);
                     connected=false;
                     return ProtoUtils.createLogoutResponse();

                 } catch (ServicesException | RemoteException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case AddBuyer:{
                 System.out.println("AddBuyerRequest ...");
                 Buyer buyer = ProtoUtils.getBuyer(request);
                 try {
                     Buyer buyer1 = server.addBuyer(buyer);
                     return ProtoUtils.createAddBuyerResponse(buyer1);
                 } catch (ServicesException | RemoteException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case BuyTicket:{
                 System.out.println("BuyTicketRequest ...");
                 ArtisticRepresentation ar = ProtoUtils.getArtisticRepresentation(request);
                 try {
                     ArtisticRepresentation artisticRepresentation1 = server.updateAR(ar, this);
                     return ProtoUtils.createBuyTicketResponse(artisticRepresentation1);
                 } catch (ServicesException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 } catch (SQLException | RemoteException throwables) {
                     throwables.printStackTrace();
                 }
             }
             case GetAllAR:{System.out.println("GetAllARRequest ...");
                 try {
                     Iterable<ArtisticRepresentation> ars = server.getAllAR();
                     return ProtoUtils.createGetAllARResponse(ars);
                 } catch (ServicesException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 } catch (SQLException | RemoteException throwables) {
                     throwables.printStackTrace();
                 }
             }
             case GetAllARFromDate:{
                 ArtisticRepresentation ar = ProtoUtils.getArtisticRepresentation(request);
                 try {
                     Iterable<ArtisticRepresentation> ars = server.getAllARFromDate(ar.getData());
                     return ProtoUtils.createGetAllARFromDateResponse(ars);
                 } catch (ServicesException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 } catch (SQLException | RemoteException throwables) {
                     throwables.printStackTrace();
                 }
             }
             case FindSellerByUsername:{
                 Seller seller = ProtoUtils.getSeller(request);
                 try {
                     Seller seller1 = server.findSellerByUsername(seller);
                     return ProtoUtils.createFindSellerByUsernameResponse(seller1);
                 } catch (ServicesException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 } catch (RemoteException throwables) {
                     throwables.printStackTrace();
                 }
             }



         }
         return response;
     }

     private void sendResponse(TSProtobufs.TSResponse response) throws IOException{
         System.out.println("sending response "+response);
         response.writeDelimitedTo(output);
         //output.writeObject(response);
         output.flush();
     }
}

File: ticketProject7-CrossPlatform/TS-crossp/protobuf3/ProtoTSWorker.cs
using System;
using System.Threading;
using System.Net.Sockets;

using Chat.Protocol;
using Google.Protobuf;
using services;
using System.Collections.Generic;

namespace protobuf
{
	public class ProtoTSWorker : ITicketSystemObserver
	{
		private ITicketSystemServices server;
		private TcpClient connection;

		private NetworkStream stream;
		private volatile bool connected;
		public ProtoTSWorker(ITicketSystemServices server, TcpClient connection)
		{
			this.server = server;
			this.connection = connection;
			try
			{

				stream = connection.GetStream();
				connected = true;
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}

		public virtual void run()
		{
			while (connected)
			{
				try
				{

					TSRequest request = TSRequest.Parser.ParseDelimitedFrom(stream);
					TSResponse response = handleRequest(request);
					if (response != null)
					{
						sendResponse(response);
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}

				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
			}
			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error " + e);
			}
		}

		public void boughtTicket(model.ArtisticRepresentation artisticRepresentation)
		{
			Console.WriteLine("Someone bought tickets " + artisticRepresentation);
			try
			{
				sendResponse(ProtoUtils.createBoughtTicketResponse(artisticRepresentation));
			}
			catch (Exception e)
			{
				throw new ServicesException("Sending error: " + e);
			}
		}

		private TSResponse handleRequest(TSRequest request)
		{
			TSResponse response = null;
			TSRequest.Types.Type reqType = request.Type;
			switch (reqType)
			{
				case TSRequest.Types.Type.Login:
					{
						Console.WriteLine("Login request ...");
						model.Seller seller = ProtoUtils.getSeller(request);
						try
						{
							
							lock (server)
							{
								model.Seller seller1;
								seller1 = server.login(seller, this);
								return ProtoUtils.createLoginResponse(seller1);
							}
							
						}
						catch (ServicesException e)
						{
							connected = false;
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.Logout:
					{
						Console.WriteLine("Logout request");
						model.Seller seller = ProtoUtils.getSeller(request);
						try
						{
							lock (server)
							{

								server.logout(seller, this);
							}
							connected = false;
							return ProtoUtils.createLogoutResponse();

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.AddBuyer:
					{
						Console.WriteLine("AddBuyer request");
						model.Buyer buyer = ProtoUtils.getBuyer(request);
						try
						{
							model.Buyer buyer1;
							lock (server)
							{
								buyer1 = server.addBuyer(buyer);
							}
							return ProtoUtils.createAddBuyerResponse(buyer1);

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.BuyTicket:
					{
						Console.WriteLine("BuyTicket request");
						model.ArtisticRepresentation artisticRepresentation = ProtoUtils.getArtisticRepresentation(request);
						try
						{
							model.ArtisticRepresentation ar1;
							lock (server)
							{
								ar1 = server.updateAR(artisticRepresentation, this);
							}
							return ProtoUtils.createBuyTicketResponse(ar1);

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.GetAllAr:
					{
						Console.WriteLine("GetAllAR request");
						try
						{
							IEnumerable<model.ArtisticRepresentation> ars;
							lock (server)
							{
								ars = server.getAllAR();
							}
							return ProtoUtils.createGetAllARResponse(ars);

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.GetAllArfromDate:
					{
						Console.WriteLine("GetAllARFromDate request");
						model.ArtisticRepresentation artisticRepresentation = ProtoUtils.getArtisticRepresentation(request);
						try
						{
							IEnumerable<model.ArtisticRepresentation> ars;
							lock (server)
							{
								ars = server.getAllARFromDate(artisticRepresentation.Data);
							}
							return ProtoUtils.createGetAllARFromDateResponse(ars);

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case TSRequest.Types.Type.FindSellerByUsername:
					{
						Console.WriteLine("FindSellerByUsername request");
						model.Seller seller = ProtoUtils.getSeller(request);
						try
						{
							model.Seller seller1;
							lock (server)
							{
								seller1 = server.findSellerByUsername(seller);
							}
							return ProtoUtils.createFindSellerByUsernameResponse(seller1);

						}
						catch (ServicesException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}



			}
			return response;
		}

		private void sendResponse(TSResponse response)
		{
			Console.WriteLine("sending response " + response);
			lock (stream)
			{
				response.WriteDelimitedTo(stream);
				stream.Flush();
			}

		}
	}



}
